internal/cache/lru/lrucache_test.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lru_test
+
+import (
+	"errors"
+	"sync"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	. "github.com/jacobsa/ogletest"
+)
+
+func TestCache(t *testing.T) { RunTests(t) }
+
+////////////////////////////////////////////////////////////////////////
+// Invariant-checking cache
+////////////////////////////////////////////////////////////////////////
+
+type invariantsCache struct {
+	Wrapped lru.Cache
+}
+
+func (c *invariantsCache) Insert(key string, value lru.ValueType) ([]lru.ValueType, error) {
+	c.Wrapped.CheckInvariants()
+	defer c.Wrapped.CheckInvariants()
+
+	return c.Wrapped.Insert(key, value)
+}
+
+func (c *invariantsCache) Erase(key string) lru.ValueType {
+	c.Wrapped.CheckInvariants()
+	defer c.Wrapped.CheckInvariants()
+
+	return c.Wrapped.Erase(key)
+}
+
+func (c *invariantsCache) LookUp(key string) lru.ValueType {
+	c.Wrapped.CheckInvariants()
+	defer c.Wrapped.CheckInvariants()
+
+	return c.Wrapped.LookUp(key)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Boilerplate
+////////////////////////////////////////////////////////////////////////
+
+const MaxSize = 50
+
+type CacheTest struct {
+	cache invariantsCache
+}
+
+func init() { RegisterTestSuite(&CacheTest{}) }
+
+func (t *CacheTest) SetUp(*TestInfo) {
+	t.cache.Wrapped = lru.New(MaxSize)
+}
+
+type testData struct {
+	Value    int64
+	DataSize uint64
+}
+
+func (td testData) Size() uint64 {
+	return td.DataSize
+}
+
+// insertAndAssert inserts the given key,value in the cache and assert based on
+// the expected eviction and error.
+func (t *CacheTest) insertAndAssert(key string, val lru.ValueType, expectedEviction int, evictedValues []int64, expectedError error) {
```

Comment:
can we skip using expectedEviction as length can be determined using evictedValues?